### ISSUE TITLE
[HOLD for payment 2023-10-13] [$500] Web - emoji picker hides on resize window

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#21560]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to any chat
2. Open emoji picker
3. Resize window (<800)
4. Verify its still open
5. Resize the window to the wider view

## Expected Result:
Emoji picker should be open

## Actual Result:
Emoji picker hides

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.30-5
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/2f5def1a-4ab0-4b3a-8240-d1535260159b

https://github.com/Expensify/App/assets/93399543/26dbee70-cfa6-4dde-9a1b-8058eb9eec6e

**Expensify/Expensify Issue URL:**
**Issue reported by:** @gadhiyamanan
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1687410686033509

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01004df42b8a89b793</li>
        <li>Upwork Job ID: 1674401243450486784</li>
        <li>Last Price Increase: 2023-09-24</li>
<li>Automatic offers: </li>
<ul>
<li>gadhiyamanan | Reporter | 26809113</li>
<li>shubham1206agra | Contributor | 26982232</li>
<li>abdulrahuman5196 | Contributor | 26995658</li>
</ul></li>
    </ul>
</details>

**Comment by jeet-dhandha** [comment]
### Proposal
### Please re-state the problem that we are trying to solve in this issue.
- When resizing the browser window, the emoji picker closes. When moving from small width to large width and vice versa.

### What is the root cause of that problem?
- The cause is that the emoji picker unmounts when the window is resized, this happens due to the `PopoverWithMeasuredContent` used in `EmojiPicker`.
- The isContentMeasured state is set to false when the window is resized, which causes the emoji picker to unmount.

### What changes do you think we should make in order to solve the problem?
- We can add a local variable in `EmojiPickerAction` file like `activeType` and a function like `setActiveType` to set the active type and `getActiveType` to get the active type.
- The activeType could vary depending on use case like `COMPOSE` for emoji button from composer input and if from AddReactionBubble which is down at chat can have `REACTION_BUBBLE_${reportActionId}`. Now whenever we call showEmojiPicker from any of this places we pass actionType and set it in the file. 
- Thus after resizing the window the Composer Emoji Button or The Reaction Bubble re-mounts and then we can check if `getActiveType()` is same then we call their own `showEmojiPicker` function.

<details>
<summary> Code Changes Explaination: </summary>

- We need to add a local variable in `EmojiPickerAction` file like `activeType` and a function like `setActiveType` to set the active type and `getActiveType` to get the active type.
- Now we set `activeType` in `showEmojiPicker` to provided type from params and also set it to `null` in `hideEmojiPicker` function.
- We also need to set `activeType` to `null` in EmojiPicker's `hideEmojiPicker` function.
- Now we need to add a `useEffect` in `EmojiPickerButton` to check if the `activeType` is 'COMPOSE' then call `showEmojiPicker` function. (Dependencies of useEffect will be empty as we only want to call it once)
- Now we need to add a `useEffect` in `AddReactionBubble` to check if the `activeType` is `ADD_REACTION_BUBBLE_${props.reportAction.reportActionID}` then call the `showEmojiPicker` function. (Dependencies of useEffect will be empty as we only want to call it once)


</details>

Note: I am still not able to fix the picker closing issue for `MiniQuickEmojiReaction`. (Will try to fix that.)

<details>
<summary> Fixed Video </summary>



https://github.com/Expensify/App/assets/78416198/f7b2616c-662b-44f4-b2ef-77fc877c8d9f


</details>

### What alternative solutions did you explore? (Optional)
- N/A

**Comment by CortneyOfstad** [comment]
Was able to recreate, so going to get eyes on the two proposals above

![2023-06-29_07-53-50 (1)](https://github.com/Expensify/App/assets/10926327/070dddbc-4277-4155-be31-b26fb462c03a)

**Comment by abdulrahuman5196** [comment]
@CortneyOfstad  Are we going to be working on this resize window issue? I thought we are not prioritizing these kind of window resizing bugs? Could you kindly confirm if this issue is something we want to fix before reviewing the proposals?

**Comment by CortneyOfstad** [comment]
@abdulrahuman5196 To be honest, I have been out the last two months, so wasn't sure about this change. But no worries. Can you reference where the conversation was had about not prioritizing these issues, so I can take a look?

**Comment by abdulrahuman5196** [comment]
Sorry @CortneyOfstad I thought this as a different issue. Apologies. This is due to the recent navigation refractor changes and the effect of mounting i assume.

**Comment by abdulrahuman5196** [comment]
@mountiny I assume this behaviour is caused due to the navigation refactors and the mounting/unmounting when we switch to smaller screens. Correct me if wrong.

What is the current expectation of popovers? Do we expect it to close when resize the window to smaller ones? (I think its fine to close)
Previous working was emoji popover wouldn't close and transform. I wanted to confirm the expectation on this before reviewing the proposals.

**Comment by mountiny** [comment]
I was ooo this week. We got couple of problems raised which are tied to the fact the navigators unmount and mount again when resizing.

In case of this one i think we should close this and do nothing in this issue. We will be discussing with SWM if there is a way to find an updated way to handle the navigators so we do not change between them which would be the root cause solution rather than solving all the symptoms of the updated architecture.

cc @adamgrzybowski @WoLewicki

**Comment by adamgrzybowski** [comment]
Currently working on the solution for resizing without remounting navigators.

**Comment by jeet-dhandha** [comment]
Same issue on this https://github.com/Expensify/App/issues/22196

**Comment by mountiny** [comment]
Removing the Help Wanted so people dont spend time on the proposals. I believe this should be fixed once Adam implements the change to how resize is implemented.

**Comment by mountiny** [comment]
@adamgrzybowski is still owrking on the styling solution without remounting

**Comment by mountiny** [comment]
This is on hold and can be weekly

**Comment by mountiny** [comment]
On hold

**Comment by CortneyOfstad** [comment]
No update yet 👍

**Comment by CortneyOfstad** [comment]
Heading OoO until 8/14 so re-assigning BZ to keep an eye on it while I'm out 👍

**Comment by mountiny** [comment]
Still on hold makin gthi monthly

**Comment by shubham1206agra** [comment]
This is fixed partially (from large to small)
But problem persists from going small to large

**Comment by mountiny** [comment]
Thanks @shubham1206agra updated the instructios

**Comment by shubham1206agra** [comment]
@mountiny I have found the problem and will post a solution tomorrow.
I need to check for regression in my solution.

**Comment by shubham1206agra** [comment]
Proof

https://github.com/Expensify/App/assets/58412969/b2de6616-e5e0-4e15-8107-ed826536eae2

**Comment by abdulrahuman5196** [comment]
Thank you @shubham1206agra 

Just to note status, We are waiting on proposal on this issue

**Comment by shubham1206agra** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.

Web - emoji picker hides on resize window

### What is the root cause of that problem?

In

https://github.com/Expensify/App/blob/ebcc8d045695871d8189a6335c6cdfeede657deb/src/components/PopoverProvider/index.js#L105-L107

We are checking for equality for ref instead of content cause multiple Popovers may have the same content (for example - EmojiPicker in different places).

But here 

https://github.com/Expensify/App/blob/ebcc8d045695871d8189a6335c6cdfeede657deb/src/components/Popover/index.js#L30-L33

We are unmounting the `PopoverWithoutOverlay` when going to the small screen width config and remounting when going out of the small screen width config. This will destroy the ref and re-create in this cycle. This causes the equality to become false, thus closing the popover.

### What changes do you think we should make in order to solve the problem?

Take the ref out of 

https://github.com/Expensify/App/blob/ebcc8d045695871d8189a6335c6cdfeede657deb/src/components/PopoverWithoutOverlay/index.js#L13

to the 

https://github.com/Expensify/App/blob/ebcc8d045695871d8189a6335c6cdfeede657deb/src/components/Popover/index.js#L13

and pass it as a prop in 

https://github.com/Expensify/App/blob/ebcc8d045695871d8189a6335c6cdfeede657deb/src/components/Popover/index.js#L30-L33

like this 

```js
createPortal(<PopoverWithoutOverlay {...props} withoutOverlayRef={refWithoutOverlay} />, document.body);
```

### What alternative solutions did you explore? (Optional)

We can disable the withoutOverlay, but this is not recommended.

**Comment by mallenexpensify** [comment]
@CortneyOfstad  I'm off this week, can you please keep 👀 on this then I'll snag it back on Monday?  Thx

**Comment by CortneyOfstad** [comment]
@mallenexpensify of course! No worries!

**Comment by CortneyOfstad** [comment]
@abdulrahuman5196 Thoughts on the proposal above?

**Comment by mountiny** [comment]
@abdulrahuman5196 lets triage the proposals now please

**Comment by mallenexpensify** [comment]
Thanks @CortneyOfstad for 👀, unassigning.

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by abdulrahuman5196** [comment]
@shubham1206agra The solution is working.
 But I am not sure if the root cause is correct. I tried to put console logs and the popover is not closing at the mentioned pointer.

https://github.com/Expensify/App/blob/ebcc8d045695871d8189a6335c6cdfeede657deb/src/components/PopoverProvider/index.js#L105-L107

>We are checking for equality for ref instead of content cause multiple Popovers may have the same content (for example - EmojiPicker in different places).

**Comment by shubham1206agra** [comment]
@abdulrahuman5196 it's really hard to check using console.log cause ref gets updated quickly

**Comment by abdulrahuman5196** [comment]
> @abdulrahuman5196 it's really hard to check using console.log cause ref gets updated quickly

@shubham1206agra We have to confirm on the root cause. If we don't confirm the root cause, for worst case the code be fix which could break later.

Or could you let me know how you verified and arrived at the root cause so that I could try to understand the same.

**Comment by shubham1206agra** [comment]
Just put a trace on hideModal method. Maybe that will help

**Comment by mountiny** [comment]
@abdulrahuman5196 how is this one looking?

**Comment by mountiny** [comment]
Bump @abdulrahuman5196

**Comment by shubham1206agra** [comment]
Friendly bump @abdulrahuman5196

**Comment by mountiny** [comment]
asking for a proposal review in slack

**Comment by abdulrahuman5196** [comment]
Just an update. On last check, the solution was working fine and theoritically correct but hard to verify the root cause. I will try today again on root cause and post update accordingly.

**Comment by abdulrahuman5196** [comment]
Reviewing now again.

**Comment by abdulrahuman5196** [comment]
@shubham1206agra 's proposal here https://github.com/Expensify/App/issues/21560#issuecomment-1705301225 looks good and works well. I was able to re-confirm the root cause as well atlast.

🎀 👀 🎀
C+ Reviewed

cc: @mountiny

**Comment by shubham1206agra** [comment]
@mountiny Please assign here

**Comment by abdulrahuman5196** [comment]
Pending on @mountiny 's review here

**Comment by mountiny** [comment]
Assigned, thanks!

**Comment by abdulrahuman5196** [comment]
@mountiny Kindly assign me as well C+ here

**Comment by abdulrahuman5196** [comment]
>The PR that introduced the bug has been identified. Link to the PR:
The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

IMO not a regression from a single change. This is something which wasn't working after navigation refractors. 

>Determine if we should create a regression test for this bug.

Yes.

>If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.


1)  Go to any chat
2) Open emoji picker
3) Resize window (<800)
4) Resize the window to a wider view

@mallenexpensify

**Comment by mountiny** [comment]
Lets make that regression test

**Comment by mountiny** [comment]
Issue created here https://github.com/Expensify/Expensify/issues/326639

**Comment by mallenexpensify** [comment]
Thanks @mountiny ! 

Issue reporter: @gadhiyamanan due $250 via Upwork
Contributor: @shubham1206agra paid $750 via Upwork, inc. urgency bonus. 
Contributor+: @abdulrahuman5196 paid $750 via Upwork, inc. urgency bonus.

@gadhiyamanan, I needed to create a new job to pay you, can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~0105078349f662f2d1

**Comment by gadhiyamanan** [comment]
@mallenexpensify offer accepted, thanks!

**Comment by mallenexpensify** [comment]
Thanks @gadhiyamanan 
I updated my comment above to denote you've been paid!
